### PR TITLE
Add grafana deployment.

### PR DESCRIPTION
**Pull request by kwoodson**
Add necessary files to create a grafana deployment.

### CODE REVIEW

File: sre/grafana/route.yaml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+---
+apiVersion: route.openshift.io/v1
+kind: Route
+metadata:
+  name: grafana
+  namespace: "grafana"
+spec:
+  host: "grafana.sre.osadev.cloud"
```
Comment (mjudeikis) [inline sre/grafana/route.yaml]: remove this, for now, let it generate

File: sre/grafana/deployment.yaml [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        - -cookie-secret-file=/etc/proxy/secrets/session_secret
+        - -skip-auth-regex=^/metrics,/api/datasources,/api/dashboards
+        volumeMounts:
+        - mountPath: /etc/tls/private
+          name: grafana-tls
+        - mountPath: /etc/proxy/secrets
+          name: secrets
+
+      - name: grafana
+        image: ${IMAGE_GRAFANA}
```
Comment (mjudeikis) [inline sre/grafana/deployment.yaml]: replace

File: Makefile [hunk 3]
```diff
@@ -12,4 +12,10 @@ setup-prometheus:
 	oc create secret generic azure-master --from-literal=azure_client_id=${AZURE_CLIENT_ID} --from-literal=azure_client_secret=${AZURE_CLIENT_SECRET} --from-literal=azure_tenant_id=${AZURE_TENANT_ID} --from-literal=azure_subscription_id=${AZURE_SUBSCRIPTION_ID} -o yaml --dry-run | oc apply -n prometheus -f -
 	oc create secret generic alertmanager --from-literal=alerts_slack_webhook_url=${ALERTS_SLACK_WEBHOOK_URL} --from-literal=alerts_slack_channel=${ALERTS_SLACK_CHANNEL}  -o yaml --dry-run | oc apply -n prometheus -f -
 
-.PHONY: setup-prometheus build-all
+setup-grafana:
+	oc apply -f sre/grafana/
```
Comment (mjudeikis) [inline Makefile]: `s/grafana/grafana-deployment/`

File: Makefile [hunk 4]
```diff
@@ -12,4 +12,10 @@ setup-prometheus:
...
+	oc get secret -n grafana grafana-proxy >/dev/null || oc create secret generic grafana-proxy --from-literal='session_secret=${SESSION_SECRET}' | oc apply -n grafana -f -
```
Comment (mjudeikis) [inline Makefile]: maybe its worth generating with session secret using bash? just use the same format as default session secret

File: sre/grafana/deployment.yaml [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+  template:
+    metadata:
+      labels:
+        app: grafana
+      name: grafana
+    spec:
+      serviceAccountName: grafana
+      containers:
+      - name: oauth-proxy
+        image: ${IMAGE_PROXY}
```
Comment (mjudeikis) [inline sre/grafana/deployment.yaml]: replace to image

File: sre/grafana/deployment.yaml [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+      - name: grafanaconfig
+        configMap:
+          name: grafana-config
+      - name: secrets
+        secret:
+          secretName: grafana-proxy
+      - name: grafana-tls
+        secret:
+          secretName: grafana-tls
+      - emptyDir: {}
```
Comment (mjudeikis) [inline sre/grafana/deployment.yaml]: add pvc template for this so we have persistence

File: Makefile (hunk 4 above)
Comment (mjudeikis) [reply Makefile]: I think just lets do oc apply and ignore the check if one exist. We dont do this for Prometheus now too

Comment (kwoodson) [reply Makefile]: If the secret exists it fails the make.

File: sre/grafana/deployment.yaml (hunk 6 above)
Comment (kwoodson) [reply sre/grafana/deployment.yaml]: Let me add that.

File: Makefile (hunk 3 above)
Comment (mjudeikis) [reply Makefile]: + add namespace object  and same structure as Prometheus has

File: Makefile [hunk 7]
```diff
@@ -12,4 +12,18 @@ setup-prometheus:
...
+setup-grafana:
+	# requires that OSADMIN_PW bet set
+	# export OSADMIN_PW="$(hack/config.sh get-config kwoodsontest35 | jq .config.AdminPasswd | tr -d '"')"
+	oc apply -Rf sre/grafana-deployment/openshift/Namespace/grafana.yaml
+	oc apply -Rf sre/grafana-deployment/openshift/
+	oc get secret -n grafana grafana-proxy >/dev/null 2>&1|| oc create -n grafana secret generic grafana-proxy --from-literal="session_secret=123456"
+	oc rollout status deployment/grafana -n grafana
+	oc adm policy add-role-to-user view -z grafana -n openshift-metrics
+
+setup-grafana-data:
```
Comment (mjudeikis) [inline Makefile]: Add to phony

**Comment by mjudeikis** [comment]
I think this s fine for first iteration. Next thing we need to think what to do with all these secret and config, dashboard generation. 
Thinking of packaging all this to its own small operator which would not lifecycle the components for now, but just updates the configs, secrets, deploy yaml files.

**Comment by mjudeikis** [comment]
Having these now we need to start writing prometheus alert rules, and grafana dashboards. Dashboards second, rules first.

File: Makefile [hunk 8]
```diff
@@ -12,4 +12,18 @@ setup-prometheus:
 	oc create secret generic azure-master --from-literal=azure_client_id=${AZURE_CLIENT_ID} --from-literal=azure_client_secret=${AZURE_CLIENT_SECRET} --from-literal=azure_tenant_id=${AZURE_TENANT_ID} --from-literal=azure_subscription_id=${AZURE_SUBSCRIPTION_ID} -o yaml --dry-run | oc apply -n prometheus -f -
 	oc create secret generic alertmanager --from-literal=alerts_slack_webhook_url=${ALERTS_SLACK_WEBHOOK_URL} --from-literal=alerts_slack_channel=${ALERTS_SLACK_CHANNEL}  -o yaml --dry-run | oc apply -n prometheus -f -
 
-.PHONY: setup-prometheus build-all
+setup-grafana:
+	# requires that OSADMIN_PW bet set
+	# export OSADMIN_PW="$(hack/config.sh get-config kwoodsontest35 | jq .config.AdminPasswd | tr -d '"')"
+	oc apply -Rf sre/grafana-deployment/openshift/Namespace/grafana.yaml
+	oc apply -Rf sre/grafana-deployment/openshift/
+	oc get secret -n grafana grafana-proxy >/dev/null 2>&1|| oc create -n grafana secret generic grafana-proxy --from-literal="session_secret=123456"
```
Comment (kwoodson) [inline Makefile]: @mjudeikis, any suggestions on how to dynamically generate this secret?  I hard coded it to make testing much easier.

File: Makefile (hunk 7 above)
Comment (kwoodson) [reply Makefile]: Added.

File: Makefile (hunk 8 above)
Comment (mjudeikis) [reply Makefile]: Default we use in `openshift-azure` is 32 lenght alfanumerical string.
`$(cat /dev/urandom | tr -dc 'a-zA-Z0-9' | fold -w 32 | head -n 1)` should do the job until we move generation to code

Comment (jim-minter) [reply Makefile]: tr -dc a-zA-Z0-9 </dev/urandom | head -c32

Comment (mjudeikis) [reply Makefile]: Even better :)

File: Makefile [hunk 9]
```diff
@@ -12,4 +12,17 @@ setup-prometheus:
...
+	oc rollout status deployment/grafana -n grafana
+	oc adm policy add-role-to-user view -z grafana -n openshift-metrics
```
Comment (jim-minter) [inline Makefile]: ideally, save the rolebinding as a yaml file in this repo

File: sre/grafana-deployment/openshift/PersistentVolumeClaim/grafana.yaml [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+  labels:
+    app: grafana
+  name: grafana
+  namespace: grafana
+spec:
+  accessModes:
+  - ReadWriteOnce
+  resources:
+    requests:
+      storage: 128Gi
```
Comment (jim-minter) [inline sre/grafana-deployment/openshift/PersistentVolumeClaim/grafana.yaml]: so big?

File: Makefile [hunk 11]
```diff
@@ -12,4 +12,17 @@ setup-prometheus:
...
+
+setup-grafana-data:
+	./sre/grafana-deployment/setup.sh
```
Comment (jim-minter) [inline Makefile]: file doesn't exist in repo

File: sre/grafana-deployment/openshift/PersistentVolumeClaim/grafana.yaml (hunk 10 above)
Comment (mjudeikis) [reply sre/grafana-deployment/openshift/PersistentVolumeClaim/grafana.yaml]: 32?

Comment (jim-minter) [reply sre/grafana-deployment/openshift/PersistentVolumeClaim/grafana.yaml]: what does grafana actually persist, out of interest?

Comment (mjudeikis) [reply sre/grafana-deployment/openshift/PersistentVolumeClaim/grafana.yaml]: Users, Dashboards, and it uses sqlite3 internally

File: Makefile (hunk 8 above)
Comment (kwoodson) [reply Makefile]: I had something but it was more like @mjudeikis: 
```
head -c 500 /dev/urandom | tr -dc 'a-zA-Z0-9~!@#$%^&*_-' | fold -w 40 | head -n 1
```

The other piece missing was to do `$$(` in make files.  Thanks for the help.

File: sre/grafana-deployment/openshift/PersistentVolumeClaim/grafana.yaml (hunk 10 above)
Comment (kwoodson) [reply sre/grafana-deployment/openshift/PersistentVolumeClaim/grafana.yaml]: Updated to 32Gi.

File: Makefile (hunk 11 above)
Comment (kwoodson) [reply Makefile]: Renamed and added. Thanks.

File: Makefile (hunk 9 above)
Comment (kwoodson) [reply Makefile]: Moved.

**Comment by kwoodson** [comment]
Squashed and removed [wip].

@jim-minter @mjudeikis

**Comment by mjudeikis** [comment]
lgtm. @jim-minter  if you happy, merge and iterate from it?

**Pull request merged**